File: app/src/main/java/com/michaelfotiadis/locationmanagerviewer/data/datastore/nmea/NougatNmeaListenerManager.java
package com.michaelfotiadis.locationmanagerviewer.data.datastore.nmea;

import android.annotation.SuppressLint;
import android.location.LocationManager;
import android.location.OnNmeaMessageListener;
import android.os.Build;

import androidx.annotation.RequiresApi;

import java.util.regex.Pattern;

public class NougatNmeaListenerManager implements NmeaListenerManager {

    private final LocationManager locationManager;
    private Listener listener;

    public NougatNmeaListenerManager(final LocationManager locationManager) {
        this.locationManager = locationManager;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    @Override
    public void register(final Callback callback) {
        listener = new Listener(callback);
        locationManager.addNmeaListener(listener);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void unregister() {
        if (listener != null) {
            locationManager.removeNmeaListener(listener);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static class Listener implements OnNmeaMessageListener {

        private final Pattern PATTERN = Pattern.compile("$");

        private final Callback callback;

        public Listener(final Callback callback) {
            this.callback = callback;
        }

        @Override
        public void onNmeaMessage(final String message, final long timestamp) {
            for (final String data : PATTERN.split(message)) {
                if (data.length() > 2) {
                    callback.onNmeaMessageReceived(data);
                }
            }
        }
    }

}

File: app/src/main/java/com/michaelfotiadis/locationmanagerviewer/data/datastore/nmea/NougatNmeaListenerManager.kt
package com.michaelfotiadis.locationmanagerviewer.data.datastore.nmea

import android.annotation.SuppressLint
import android.location.LocationManager
import android.location.OnNmeaMessageListener
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.annotation.RequiresApi
import java.util.regex.Pattern

class NougatNmeaListenerManager(
    private val locationManager: LocationManager
) : NmeaListenerManager {

    private var listener: Listener? = null

    @RequiresApi(api = Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    override fun register(callback: NmeaListenerManager.Callback) {
        listener = Listener(callback).apply {
            locationManager.addNmeaListener(this, Handler(Looper.getMainLooper()))
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun unregister() {
        if (listener != null) {
            locationManager.removeNmeaListener(listener!!)
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    class Listener(private val callback: NmeaListenerManager.Callback) : OnNmeaMessageListener {
        override fun onNmeaMessage(message: String, timestamp: Long) {
            for (data in PATTERN.split(message)) {
                if (data.length > 2) {
                    callback.onNmeaMessageReceived(data)
                }
            }
        }
    }

    private companion object {
        val PATTERN: Pattern = Pattern.compile("$")
    }
}